Replace Try::Tiny with Feature::Compat::Try

b10n1k opened the pull request:
In general straight forward refactoring.
However there are some points which needs some additional attention:
- when return try..catch
- the return value (either $_, or other complex structure)
- when return from try block
  
Two modules which has `return` inside _try_ have been left out and
they will be worked in another PR to enphasize to the quality and
proper refactoring.
    


https://progress.opensuse.org/issues/176862

okurz commented on lib/OpenQA/LiveHandler/Controller/LiveViewHandler.pm:
```diff
@@ -197,7 +197,7 @@ sub handle_message_from_java_script {
         $json = decode_json($msg);
     }
-    catch {
+    catch ($error) {
```
The common style is `catch ($e)` same as we use in os-autoinst. No reason to be different just for the sake of it.

Same for other occurences later on

b10n1k replied:
not for the sake to be different. 
- IMO it is more clear
- can reused in other places where $_ is stored in $error

b10n1k replied:
i will update it

okurz commented on lib/OpenQA/Scheduler/Model/Jobs.pm:
```diff
@@ -178,9 +178,9 @@ sub schedule ($self) {
         }
-        catch {
+        catch ($error) {
             log_debug("Failed to retrieve worker ($worker_id) in the DB, reason: $_");    # uncoverable statement
```
Pretty sure this must be

```suggestion
            log_debug("Failed to retrieve worker ($worker_id) in the DB, reason: $e");    # uncoverable statement
```

Same for other occurences later on

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:
```diff
@@ -568,9 +567,9 @@ sub update_status ($self) {
+    catch ($error) {
         # uncoverable statement
-        my $error_message = $_;
+        my $error_message = $error; #TOD refactor
```
?

b10n1k replied:
there are parts i want to come back. they look that can modified accordingly

Wabri replied:
I think `TOD` is a typo of `TODO` and since this is a draft pr I dont see problem on leaving it there while work in progress

perlpunk commented on lib/OpenQA/Scheduler/Model/Jobs.pm:
```diff
@@ -178,9 +178,9 @@ sub schedule ($self) {
         }
-        catch {
+        catch ($error) {
             log_debug("Failed to retrieve worker ($worker_id) in the DB, reason: $_");    # uncoverable statement
```
```suggestion
            log_debug("Failed to retrieve worker ($worker_id) in the DB, reason: $e");    # uncoverable statement
```

perlpunk replied:
And the same for all other instances. Try::Tiny uses `$_`, and you can replace that with the named variable.

perlpunk commented on lib/OpenQA/Schema/Result/ScheduledProducts.pm:
```diff
@@ -385,7 +387,7 @@ sub _schedule_iso {
             }
-            catch {
+            catch ($error) {
                 my $error = shift;
```
```suggestion
```

b10n1k commented:
@okurz @perlpunk https://progress.opensuse.org/issues/176862#note-7. please be patient. Thats why I used the labels.

perlpunk commented:
> @okurz @perlpunk https://progress.opensuse.org/issues/176862#note-7. please be patient. Thats why I used the labels.

sorry. It sounded like you were unsure about some things and could use suggestions.

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -182,11 +182,14 @@ sub create {
-
-    if ($error) {
-        return $self->render(json => {error => $error}, status => 400);
+    try { 
+        warn ">>>>>>>>>another hit";
+        $id = $self->schema->resultset($table)->create(\%entry)->id; }
+    catch ($e) {
+        $error = $e;
     }
+    warn ">>>>>>>>>>>>>another hit errro"; 
+        return $self->render(json => {error => $error}, status => 400);
```
```suggestion
    try { 
        $id = $self->schema->resultset($table)->create(\%entry)->id; }
    catch ($e) {
        return $self->render(json => {error => $e}, status => 400);
    }
```

perlpunk commented on lib/OpenQA/Schema/Result/ScheduledProducts.pm:
```diff
@@ -270,10 +272,10 @@ sub _create_jobs_in_database ($self, $jobs, $failed_job_info, $skip_chained_deps
-        catch {
+        catch ($e) {
             $schema->svp_rollback('try_create_job_from_settings');
             die $_ if $schema->is_deadlock($_);    # uncoverable statement
```
```suggestion
            die $_ if $schema->is_deadlock($e);    # uncoverable statement
```

perlpunk commented on lib/OpenQA/WebAPI.pm:
```diff
@@ -27,14 +27,17 @@ sub startup ($self) {
-        OpenQA::Setup::load_plugins($self);
+    if ($ENV{MOJO_HELP}) {
+        try {
+            my $config_result = OpenQA::Setup::read_config($self);
+            my $plugin_result = OpenQA::Setup::load_plugins($self);
+            return ($config_result, $plugin_result);
```
It's probably enough to do:
```suggestion
            OpenQA::Setup::read_config($self);
            OpenQA::Setup::load_plugins($self);
```
I think it doesn't need a return value.

perlpunk replied:
It does need a `return` statement though

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:
```diff
@@ -805,8 +804,8 @@ sub done ($self) {
-        $self->render(status => 400, json => {error => $_});
+    catch ($e) {
+        $self->render(status => 400, json => {error => $e});
     };
```
```suggestion
    }
```

perlpunk commented on lib/OpenQA/Schema/ResultSet/Assets.pm:
```diff
@@ -332,9 +332,9 @@ sub status {
         }
-        catch {
+        catch ($e) {
             log_warning("Unable to create cache file $cache_file_path: $@");    # uncoverable statement
```
```suggestion
            log_warning("Unable to create cache file $cache_file_path: $e");    # uncoverable statement
```

b10n1k replied:
thanks I was not sure about that.

b10n1k commented on lib/OpenQA/WebAPI/Controller/API/V1/JobTemplate.pm:
```diff
@@ -90,7 +90,6 @@ Returns a YAML template representing the job group(s).
 
 sub schedules {
     my $self = shift;
-
```
not done with this file. I will fix those

b10n1k commented:
latest sumbission update:
- JobTemplate.pm is not ready.
- there is an error in CI about unrelated function which somehow is affected
- notice the regex in style test

perlpunk commented:
Please add this patch to get rid of these warnings:
```
unexpected OP_CUSTOM (catch) at /usr/lib/perl5/5.26.1/B/Deparse.pm line 1667.
```
```diff
diff --git a/Makefile b/Makefile
index 801417902..374c14cfe 100644
--- a/Makefile
+++ b/Makefile
@@ -251,7 +251,7 @@ COVERDB_SUFFIX ?=
 # We use JSON::PP because there is a bug producing a (harmless) 'redefined'
 # warning when using Devel::Cover and Cpanel::JSON::XS
 # https://progress.opensuse.org/issues/90371
-COVEROPT ?= -mJSON::PP -MDevel::Cover=-select_re,'^/lib',+ignore_re,lib/perlcritic/Perl/Critic/Policy,-coverage,statement,-db,cover_db$(COVERDB_SUFFIX),
+COVEROPT ?= -mJSON::PP -MSyntax::Keyword::Try::Deparse -MDevel::Cover=-select_re,'^/lib',+ignore_re,lib/perlcritic/Perl/Critic/Policy,-coverage,statement,-db,cover_db$(COVERDB_SUFFIX),
 endif
 
 .PHONY: coverage
```

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -360,9 +361,9 @@ sub _prepare_settings {
-        catch {
-            $error = $_;
-        };
+        catch ($e) {
+            $error = $e;
+        }
         # of course stupid perl doesn't let you return directly from catch
         # as that would actually lead to readable code
         return $error if (defined $error);
```
```suggestion
        catch ($e) { return $e }
```

perlpunk commented:
> Please add this patch to get rid of these warnings:

I created a PR because we see the warnings already on master:
https://github.com/os-autoinst/openQA/pull/6206

perlpunk commented on lib/OpenQA/WebAPI.pm:
```diff
@@ -27,14 +27,16 @@ sub startup ($self) {
+        }
+        catch ($e) {
+            print("The help might be incomplete because an error occurred when loading plugins:\n$e\n");
+            return;
+        }
```
```suggestion
        }
        return;
```

okurz commented on lib/OpenQA/Schema/Result/JobModules.pm:
```diff
@@ -13,7 +13,6 @@ use Mojo::Util 'decode';
 use File::Basename qw(dirname basename);
 use File::Path 'remove_tree';
 use Cwd 'abs_path';
-use Try::Tiny;
```
please put those changes into a separate commit about removing unused imports

b10n1k commented:
I saw all the comments. But I still need to sort out one or two things.
-  the impacted api test
- now the JobTemplate.pm has duplicated code. I will thing of something better

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/JobTemplate.pm:
```diff
@@ -56,20 +56,31 @@ sub list {
-            if (my $value = $self->param('machine_name')) { $cond{'machine.name'} = $value }
-            if (my $value = $self->param('test_suite_name')) { $cond{'test_suite.name'} = $value }
+            if (my $value = $self->param('machine_name')) {
+                $cond{'machine.name'} = $value;
+            }
+            if (my $value = $self->param('test_suite_name')) {
+                $cond{'test_suite.name'} = $value;
+            }
             for my $id (qw(arch distri flavor version)) {
-                if (my $value = $self->param($id)) { $cond{"product.$id"} = $value }
+                if (my $value = $self->param($id)) {
+                    $cond{"product.$id"} = $value;
+                }
```
I would say a seperate PR. why should we have perltidy changes in this PR? it's not really related

okurz commented on lib/OpenQA/Schema.pm:
```diff
@@ -151,7 +151,10 @@ sub read_application_secrets ($self) {
 }
 
-sub is_deadlock ($self, $error) { $error =~ DEADLOCK_REGEX }
+sub is_deadlock ($self, $error) {
+    return $error =~ DEADLOCK_REGEX if defined $error;
+    return undef;
+}
```
```suggestion
sub is_deadlock ($self, $error) { defined $error ? $error =~ DEADLOCK_REGEX : undef }
```

okurz commented on lib/OpenQA/Schema.pm:
```diff
@@ -103,23 +107,25 @@ sub _try_deploy_db ($dh) {
+    }
 
     return !$version;
 }
 
 sub _try_upgrade_db ($dh) {
     my $schema = $dh->schema;
-    return 0 unless $dh->schema_version > $dh->version_storage->database_version;
+    return 0
+      unless $dh->schema_version > $dh->version_storage->database_version;
```
Those changes are not related and should be *at least* in a separate commit

okurz commented on lib/OpenQA/Schema.pm:
```diff
@@ -136,6 +142,7 @@ sub create_system_user ($self) {
 
 # read application secret from database
 sub read_application_secrets ($self) {
+
```
unrelated whitespace changes

b10n1k replied:
just wait. at least as far the `not-ready` flag in on) please

perlpunk commented on lib/OpenQA/Schema.pm:
```diff
@@ -151,9 +151,12 @@ sub read_application_secrets ($self) {
 }
 
-sub is_deadlock ($self, $error) { $error =~ DEADLOCK_REGEX }
+sub is_deadlock ($self, $error) {
+    return $error =~ DEADLOCK_REGEX if defined $error;
+    return undef;
+}
```
This is not necessary.

perlpunk commented on lib/OpenQA/Schema/Result/ScheduledProducts.pm:
```diff
@@ -270,12 +272,13 @@ sub _create_jobs_in_database ($self, $jobs, $failed_job_info, $skip_chained_deps
             $schema->svp_rollback('try_create_job_from_settings');
-            die $_ if $schema->is_deadlock($_);    # uncoverable statement
-            push @$failed_job_info, {job_name => $settings->{TEST}, error_message => $_};
+            die $e if $schema->is_deadlock($_);    # uncoverable statement
```
so no need to change `is_deadlock`

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:
```diff
@@ -568,16 +567,16 @@ sub update_status ($self) {
+    catch ($e) {
         # uncoverable statement
-        my $error_message = $_;
+        my $error_message = $e;
```
Just use `$e` below

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -182,11 +182,13 @@ sub create {
-    try { $id = $self->schema->resultset($table)->create(\%entry)->id; } catch { $error = shift; };
-
-    if ($error) {
+    try {
+        $id = $self->schema->resultset($table)->create(\%entry)->id;
+    }
```
```suggestion
    try { $id = $self->schema->resultset($table)->create(\%entry)->id }
```

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -182,11 +182,13 @@ sub create {
     my $error;
```
Isn't that variable unused now? Then remove it

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -274,10 +276,11 @@ sub update {
     }
-    catch {
+    catch ($e) {
+
```
```suggestion
```

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -182,11 +182,13 @@ sub create {
+        $id = $self->schema->resultset($table)->create(\%entry)->id;
+    }
+    catch ($e) {
         return $self->render(json => {error => $error}, status => 400);
```
```suggestion
        return $self->render(json => {error => $e}, status => 400);
```

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -354,18 +358,17 @@ sub _prepare_settings {
     my ($self, $table, $entry) = @_;
     my $validation = $self->validation;
     my $hp;
+
```
```suggestion
```

perlpunk commented on lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm:
```diff
@@ -186,15 +186,16 @@ sub create {
         }
         else {
-            $self->render(status => 500, json => {error => "Failed: $_"});
+            $self->render(status => 500, json => {error => "Failed: $e"});
+            $self->render(status => 500, json => {error => "Failed: $e"});
```
```suggestion
            $self->render(status => 500, json => {error => "Failed: $e"});
```

perlpunk commented on lib/OpenQA/WebSockets/Controller/Worker.pm:
```diff
@@ -224,10 +224,10 @@ sub _message {
             $worker->reschedule_assigned_jobs([$current_job, @unfinished_jobs]);
         }
-        catch {
+        catch ($e) {
```
The `try` block above is quite long and has three `return` statements.
Please check if this is really working as intended. Tests might pass, but maybe just because this is not covered.

perlpunk replied:
If in doubt, leave that file out for now. Before doing any refactorings.

perlpunk commented on t/deploy.t:
Shouldn't that have been part of #6211 ?

b10n1k replied:
I dont think so. this is not an unused import

b10n1k commented:
@perlpunk I should have addressed all your comments.
In additional two modules have been left out. and likely i should have left out the check as well. I will do later on or tmr.

perlpunk commented on lib/OpenQA/LiveHandler/Controller/LiveViewHandler.pm:
```diff
@@ -197,7 +197,7 @@ sub handle_message_from_java_script {
         $json = decode_json($msg);
     }
-    catch {
+    catch ($e) {
```
Leftover comma at the end of the catch block

perlpunk commented on t/deploy.t:
```diff
@@ -18,7 +18,7 @@ use OpenQA::Test::TimeLimit '20';
 use Mojo::File 'path';
 use List::Util 'min';
-use Try::Tiny;
+use Feature::Compat::Try;
```
Where is the actual change to the try/catch then?

b10n1k replied:
what do you mean `actual change`?

okurz replied:
Well, you only replaced the import but Try::Tiny and Feature::Compat::Try use different syntax. In the meantime you already changed the code by removing the semicolon but as I commented in https://github.com/os-autoinst/openQA/pull/6183#discussion_r1971094512 you should remove this import and change the code to use Test::Exception functions

perlpunk replied:
With actual change I meant, you changed the module, but not the actual `try` statement.
The error message was:
https://app.circleci.com/pipelines/github/os-autoinst/openQA/15882/workflows/87ef8c53-4203-44c1-9a58-398412fdd72f/jobs/152826
```
[17:34:55] t/deploy.t ................................................ Expected a catch {} block at t/deploy.t line 51.
```
That means Feature::Compat::Try always wants a `catch` block.

okurz commented on t/deploy.t:
```diff
@@ -48,7 +48,7 @@ my $dh = DBIx::Class::DeploymentHandler->new(
 try {
     $deployed_version = $dh->version_storage->database_version;
-};
+}
```
that's invalid syntax and needs to be

```suggestion
}
catch ($e) { }
```

However as this is part of tests we should use functions from Test::Exception

okurz commented on t/01-style.t:
```diff
@@ -22,6 +22,9 @@ is qx{git grep -I -l '^use Test::\\(Exception\\|Fatal\\)' t/**.t}, '',
 is qx{git grep -I -l '^\\(throws\\|dies\\|lives\\)_ok.*\<sub\>' t/**.t}, '',
   'Only use simplified prototyped Test::Exception functions';
 is qx{git grep -I -l 'like.*\$\@' t/**.t}, '', 'Use throws_ok instead of manual checks of exceptions';
+# https://progress.opensuse.org/issues/176862 - It should be enaled once all modules have Feature::Compat::Try
+#is qx{git grep -I -l '^use \\(Try::Tiny\\|TryCatch\\)' lib/**.pm script/** t/**.t}, '',
+#  'No Try::Tiny or TryCatch necessary, use Feature::Compat::Try and later native Perl';
```
As you didn't include all necessary changes in this PR anymore as we discussed this change should not be part of this PR but kept in a separate branch that you rebase on top

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -360,12 +358,10 @@ sub _prepare_settings {
-        # of course stupid perl doesn't let you return directly from catch
-        # as that would actually lead to readable code
-        return $error if (defined $error);
+        catch ($e) {
+            return $e;
+        }
+
```
```suggestion
        catch ($e) { return $e }
```

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -324,10 +321,11 @@ sub destroy {
     }
-    catch {
+    catch ($e) {
+
```
please remove this blank line. I don't see a need to add it

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -360,12 +358,8 @@ sub _prepare_settings {
-        # as that would actually lead to readable code
-        return $error if (defined $error);
+        catch ($e) { return $e }
+
```
remove
